src/disk_io_thread_pool.cpp

```diff
@@ -0,0 +1,200 @@
+			{
+				auto new_end = std::remove_if(m_threads.begin(), m_threads.end()
+					, [id, &l](std::thread& t)
+				{
+					if (t.get_id() == id)
+					{
+						t.detach();
+						return true;
+					}
+					return false;
+				});
+				TORRENT_ASSERT(new_end != m_threads.end());
+				m_threads.erase(new_end, m_threads.end());
+				if (m_threads.empty()) m_idle_timer.cancel();
+			}
+		}
+		return to_exit > 0;
+	}
+
+	std::thread::id disk_io_thread_pool::first_thread_id()
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_threads.empty()) return std::thread::id();
+		return m_threads.front().get_id();
+	}
+
+	void disk_io_thread_pool::job_queued(int queue_size)
+	{
+		// this check is not strictly necessary
+		// but do it to avoid aquiring the mutex in the trivial case
+		if (m_num_idle_threads >= queue_size) return;
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_abort) return;
+
+		// reduce the number of threads requested to stop if we're going to need
+		// them for these new jobs
+		int to_exit = m_threads_to_exit;
+		while (to_exit > (std::max)(0, m_num_idle_threads - queue_size) &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit
+				, (std::max)(0, m_num_idle_threads - queue_size)));
+
+		// now start threads until we either have enough to service
+		// all queued jobs without blocking or hit the max
+		for (int i = m_num_idle_threads
+			; i < queue_size && int(m_threads.size()) < m_max_threads
+			; ++i)
+		{
+			// if this is the first thread started, start the reaper timer
+			if (m_threads.empty())
+			{
+				m_idle_timer.expires_from_now(reap_idle_threads_interval);
+				m_idle_timer.async_wait([this](error_code const& ec) { reap_idle_threads(ec); });
+			}
+
+			// work keeps the io_service::run() call blocked from returning.
+			// When shutting down, it's possible that the event queue is drained
+			// before the disk_io_thread has posted its last callback. When this
+			// happens, the io_service will have a pending callback from the
+			// disk_io_thread, but the event loop is not running. this means
+			// that the event is destructed after the disk_io_thread. If the
+			// event refers to a disk buffer it will try to free it, but the
+			// buffer pool won't exist anymore, and crash. This prevents that.
+			m_threads.emplace_back(&pool_thread_interface::thread_fun
+				, &m_thread_iface, io_service::work(m_idle_timer.get_io_service()));
+		}
+	}
+
+	void disk_io_thread_pool::reap_idle_threads(error_code ec)
+	{
+		// take the minimum number of idle threads during the last
+		// sample period and request that many threads to exit
+		if (ec) return;
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_abort) return;
+		if (m_threads.size() == 0) return;
+		m_idle_timer.expires_from_now(reap_idle_threads_interval);
+		m_idle_timer.async_wait([this](error_code const& e) { reap_idle_threads(e); });
+		int min_idle = m_min_idle_threads.exchange(m_num_idle_threads);
+		if (min_idle <= 0) return;
+		stop_threads(min_idle);
```

Comment:
I guess I'm not collecting test coverage from sims, that's why this block is marked as not covered..